File: C++/boats-to-save-people.cpp
// Time:  O(nlogn)
// Space: O(n)

class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        sort(people.begin(), people.end());
        int result = 0;
        int left = 0, right = people.size() - 1;
        while (left <= right) {
            ++result;
            if (people[left] + people[right] <= limit) {
                ++left;
            }
            --right;
        }
        return result;
    }
};

File: Python/boats-to-save-people.py
# Time:  O(nlogn)
# Space: O(n)

class Solution(object):
    def numRescueBoats(self, people, limit):
        """
        :type people: List[int]
        :type limit: int
        :rtype: int
        """
        people.sort()
        result = 0
        left, right = 0, len(people)-1
        while left <= right:
            result += 1
            if people[left] + people[right] <= limit:
                left += 1
            right -= 1
        return result

